src/base/emu-i386/simx86/Makefile

```diff
@@ -12,7 +12,7 @@ EM86DIR=$(REALTOPDIR)/src/emu-i386/simx86
 EM86FLG=-Dlinux -DDOSEMU
 ifeq ($(X86_JIT),1)
 JITFILES = codegen-x86.c fp87-x86.c sigsegv.c cpatch.c trees.c
-ALL_CPPFLAGS += -mno-red-zone
+ALL_CPPFLAGS += -mno-red-zone -fomit-frame-pointer
```

Comment:
Why did you choose rbp, out of gp regs?

Reply:
But is it any better than ofs_MEMBASE?
With ofs_MEMBASE you don't ultimately
need to add. You can still use some tmp
reg and use the SIB-formed in-place additions,
similar to what you do with ebp, but w/o ebp.

Reply:
I could use e.g. rdx as the tmp reg with some effort, though it adds code to the JIT (every mem load and store then needs to add e.g. `movq ofs_MEMBASE(%rbx), %rdx`).

So it's more convenient to have it in a register. The question is then, why not rbp? Are you concerned about debugging?

The `-fomit-frame-pointer` comes from this discussion
https://github.com/dosemu2/dosemu2/pull/1844#discussion_r1051643642
I can certainly narrow the scope of it, though in any case it only really does something with `-O0`, e.g. with `-Og` it's already the default unless I missed something.

Reply:
No, debugging with frame pointers is
a dead horse today.
There are just 2 things that make me
wonder:
1. Portability (not sure if -fomit-frame-pointers
    is going to be supported everywhere)
2. I find it a bit strange that you want a separate
    reg for mem_base.
Yes, you'll get an extra moves for sure.
Less optimal jitting, obviously.
BUT. You have all the guest regs accessible
via %rbx. I am sure you could gain more
if you eg put guest's rax into rbp, or alike.
So I just wonder if mem_base is such an
important target to consume the separate
reg.
But if you think it is, then ok.

Reply:
1. well if -fomit-frame-pointer isn't supported we can always save and restore `%rbp` manually
2. This is of course a much more complicated issue, I looked at it a little bit.

We could of course keep the old method for basic memory instructions, and instead of SIB just use `%rdi`.
For stack instructions such as push and pop it's more complicated though, the 32bit seg+off needs to be recalculated for every push/pop to account for wraparound and only then added to membase. Still not impossible, as `%rdx` is available to hold membase temporarily.

Obviously memory reads/writes/rmw are very common, certainly if you include push/pop/call/ret,stringops,etc.
Guest rax is used a lot as well, whether more or less than memory, I'm not sure, would have to do some statistics.

The JITter becomes quite a bit more complex if I cache just rax though, as lots of instructions now need a special case for it, in fact on x86-64 it may be easier to put all 8 basic i386 registers (eax,ebx,ecx,edx,esp,ebp,esi,edi) in r8 to r15, as then the decoding is consistent for all 8.

So the question is then, is there something else useful to put in rbp? Cause if there isn't, might as well use it for simplicity?

Reply:
> So the question is then, is there something else useful to put in rbp?

I personally wouldn't put anything
into rbp. :) But at the end of the day
it doesn't matter.
What I still don't understand is why
use %rdi instead of SIB? I thought
you can do exactly the same SIBs
as with rbp, just with rdi?
And what stack complications do
you mean if the same code (ofs_MEMBASE)
is there right now?

Reply:
Its not like I am trying to talk you
into anything. Just trying to understand
the reasoning.

Reply:
Two issues when I try to actually do it:
1. If I put membase in %rdx temporarily, %rdx is not preserved in cpatch, operations, etc. (unlike %rdi). That means read-modify-write instructions need to load %rdx for both load and store. No big deal, just means two extra moves instead of one for those.
2. O_PUSH (and similar, O_PUSH2 etc). This one is harder, and I can't solve easily (without introducing even more reloads). Without this PR a little shortcut is used, it uses `movw %%ax,(%%esi,%%ecx,1)` where `%esi` = membase + segment base of guest %ss, and `%ecx` is the guest sp or esp. I can't do that with 64bit membase, since segment base + off needs to go in 32bit arithmetic to account for wrap around at 4G. So I need to store base+off somewhere (e.g. %rdx) and now I'm out of registers to get the second reg for the SIB for `movw %%ax,(%%???,%%rdx,1)`. All other ones are still in flight. Use `rbp` for `???` and I'm done, nicely preserved, and otherwise unused register.

Reply:
OK the last question is then, can you
just add rbp to the list of temporaries?
Then you can use it just as you do in
this PR, but you also save/restore it
when calling the jitted code, and by
that do not need to disable an FP.

By the way, why do you even need to
disable an FPs? You probably can save/restore
rbp even with your current approach?

Reply:
Or add it to clobber list to let gcc
decide whether to save/restore?

Reply:
So if rbp is allowed as a temp, but used nowhere else, might as well fix it to membase for simplicity, that's the idea here. If we find some other use for it, it can always be unfixed in the future.

As for save/restore, that's what I did in the initial patch:
https://github.com/dosemu2/dosemu2/commit/3c4dfc6e854eac074bb2622abf9a28186316e283
which we then discussed here:
https://github.com/dosemu2/dosemu2/pull/1844#discussion_r1051643642

maybe better to make all that asm code redzone-safe and thereby eliminate both `-mno-red-zone -fomit-frame-pointer`?

Reply:
> Or add it to clobber list to let gcc decide whether to save/restore?

I did put it in the clobber list here.
gcc refuses ebp in the clobber list with `-fno-omit-frame-pointer` or `-O0`.

Reply:
redzone-safe may be tricky unless you
disable redzone with an attribute on a
function that does inline asm. I don't
remember any gcc asm constraint that
says "I use stack in this asm, so please
decrement the rsp below your data".
So to me its not like the code is redzone-unsafe,
but rather that gcc is silly. But if you know
how - please do.

> As for save/restore, that's what I did in the initial patch:

I wonder why not push/pop.
But in that case you don't need fomit-frame-pointers,
right?